### PR TITLE
Fix Hovers defaults as well as some potential crashes

### LINKED ISSUE #16336: Wrong offsets for hovering units in RA/TD
We reviewers forgot to test RA and TD for regressions in #16261, both of them use `Hovers` for helis/Orca.

Those are now placed too high relative to their selection box.

We should also take a closer look at TS Harpy and Orca.

This might actually need an update rule, or we need to change the InitialHeight default to something much lower and give the TS Hover Tank a custom value instead.

### PR DESCRIPTION
**Pull request by reaperrr**
Fixes #16336.

Details are in the commit descriptions.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Render/Hovers.cs [hunk 1]
```diff
@@ -21,21 +21,23 @@ namespace OpenRA.Mods.Common.Traits.Render
 	[Desc("Changes the visual Z position periodically.")]
 	public class HoversInfo : ConditionalTraitInfo, Requires<IMoveInfo>
 	{
-		[Desc("Amount of Z axis changes in world units.")]
-		public readonly int OffsetModifier = -43;
+		[Desc("Amount of Z axis changes in world units. Must be a negative value!")]
+		public readonly WDist OffsetModifier = new WDist(-43);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/Hovers.cs]: Can we rename this to something like `BobHeight`? Our standard convention is to only use "Modifier" to mean a percentage multiplier.

**Comment by pchote** [comment]
Clamping values that are set in the yaml at runtime IMO isn't great. Can we instead implement `RulesetLoaded` and throw a `YamlException` if the values are bogus, like we do for other cases instead?

**Comment by reaperrr** [comment]
> Can we instead implement RulesetLoaded and throw a YamlException if the values are bogus?

The problem is that without the clamping, some of the bugs trigger even on values that don't look suspicious at all at first glance, because the diff between InitialHeight and OffsetModifier plays into that, too (in addition to the tick values).
Calculating on which values to throw would be a real pain, and I currently don't have the brain bandwidth and motivation to properly debug/fix/refactor the math of this trait, either.

All in all and considering how Hover Tanks worked in original TS, I'm beginning to think this trait isn't really suited to alter the visual 'center position' of hovering actors.
Things like falling to ground on EMP/Ion storm, sinking if that happens over water, and the twirling during fall all would be much easier to do if hover units actually changed their *real* altitude, via Mobile/Locomotor.
In the long run I think this trait should only handle the slight bobbing up and down, like it used to.

That said, unless you feel like taking over and reviewing + fixing the traits' math properly, it's basically 'take as-is' (except for minor fixes, if necessary) or revert the PR that introduced those issues, though that would also lose us some functionality again.

**Comment by pchote** [comment]
Ok, makes sense and I agree. Will re-review with a new focus on this as a stop-gap solution to improve the current behaviour until a future PR can move the global hover height/disabling to Mobile / Locomotor.

**Comment by pchote** [comment]
Can we please at least check:

> Must be a negative value!

> Must be higher than zero.

> Must be at least as high as RiseTicks!

Using `RulesetLoaded`?

**Comment by reaperrr** [comment]
> Using RulesetLoaded? We can then omit them from the descriptions.

Done. That actually also allowed to remove all but one clamp, so it's pretty close to your original request.

**Review by pchote — DISMISSED** [review]
LGTM. Just my comment above about `OffsetModifier`, which still stands.

**Comment by reaperrr** [comment]
Updated.

**Comment by matjaeck** [comment]
> until a future PR can move the global hover height/disabling to Mobile / Locomotor

Do we want / have an issue for this?

**Review by ghost — DISMISSED** [review]
LGTM

File: OpenRA.Mods.Common/Traits/Render/Hovers.cs [hunk 2]
```diff
@@ -21,23 +21,45 @@ namespace OpenRA.Mods.Common.Traits.Render
 	[Desc("Changes the visual Z position periodically.")]
 	public class HoversInfo : ConditionalTraitInfo, Requires<IMoveInfo>
 	{
-		[Desc("Amount of Z axis changes in world units.")]
-		public readonly int OffsetModifier = -43;
+		[Desc("Maximum visual Z axis distance relative to actual position + InitialHeight.")]
+		public readonly WDist BobDistance = new WDist(-43);
 
-		public readonly int MinHoveringAltitude = 0;
+		[Desc("Actual altitude of actor needs to be this or higher to enable hover effect.")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Render/Hovers.cs]: This sentence does not make sense.

File: OpenRA.Mods.Common/Traits/Render/Hovers.cs [hunk 3]
```diff
@@ -46,15 +68,17 @@ public class Hovers : ConditionalTrait<HoversInfo>, IRenderModifier, ITick
...
 
 		public Hovers(HoversInfo info, Actor self)
 			: base(info)
 		{
 			this.info = info;
-			this.stepPercentage = 256 / info.Ticks;
-			this.fallTickHeight = (info.InitialHeight + info.OffsetModifier) / info.FallTicks;
+			stepPercentage = 256 / info.Ticks;
+
+			// fallTickHeight must be at least 2 to avoid a DivideByZeroException and other potential problems when trait is disabled.
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Render/Hovers.cs]: Can you explain this? I don't understand how we will get a divide by zero if this value is 1.
`var fallTicks = worldVisualOffset.Z / fallTickHeight - 1;` should divide first, then subtract one.

**Review by abcdefg30 — DISMISSED** [review]
You are changing a bunch of default values as it looks. Are we fine with not having an update rule? (I actually assume yes as you labelled those default changes as bugfixes.)

**Comment by abcdefg30** [comment]
Need a rebase, too.

**Comment by pchote** [comment]
> You are changing a bunch of default values as it looks.

These are mostly restoring the original values from an earlier PR that did not provide an update rule.

File: OpenRA.Mods.Common/Traits/Render/Hovers.cs (hunk 3 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/Render/Hovers.cs]: Right, but the value still needs to be at least 1 to avaoid said DBZ Exception. Updated accordingly.

**Comment by reaperrr** [comment]
Rebased & updated.

**Review by abcdefg30 — COMMENTED** [review]
Orcas in TD still bob differently compared to the last release.

**Comment by pchote** [comment]
The only way to "fix" that is to revert #16261, which fundamentally changed the bob calculation.

**Comment by abcdefg30** [comment]
I think increasing the speed of the bobbing will make it looks similar to the release again, for example halfing to `Ticks: 6`.

**Comment by reaperrr** [comment]
> for example halfing to Ticks: 6.

Done.

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
Imho this PR has changed a bit since the initial reviews, so waiting for someone else to reconfirm their :+1: by looking if the new bobbing speeds look acceptable.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/84ae4beac99ccd644a026924c7b7ffcf7074fd03

**Pull request merged**